src/coreclr/jit/optimizebools.cpp

```diff
@@ -1771,6 +1786,100 @@ GenTree* OptBoolsDsc::optIsBoolComp(OptTestInfo* pOptTest)
     return opr1;
 }
 
+//-----------------------------------------------------------------------------
+//  optOptimizeAndConditionWithEqualityOperator: Optimize boolean when m_b1 is BBJ_COND with EQ on integers, m_b2 is
+//  BBJ_RETURN with EQ on integers and m_b3 is BBJ_RETURN of False
+//
+// Arguments:
+//      b3:    Pointer to basic block b3
+//
+//  Returns:
+//      true if boolean optimization is done and m_b1, m_b2 and m_b3 are folded into m_b1, else false.
+//
+//  Notes:
+//      m_b1, m_b2 and m_b3 of OptBoolsDsc are set on entry.
+//
+//      if B1->TargetIs(b3), it transforms
+//          B1 : brtrue(t1, B3)
+//          B2 : ret(t2)
+//          B3 : ret(0)
+//      to
+//          B1 : ret((!t1) && t2)
+//
+bool OptBoolsDsc::optOptimizeAndConditionWithEqualityOperator(BasicBlock* b3)
+{
+#if defined(TARGET_ARM) || defined(TARGET_ARM64)
+    return false;
+#endif
```

Comment:
Why this check? There should be a comment why the transformation isn't profitable on these platforms.

Reply:
It seems to conflict with if-conversion pass, I think the way this is lowered does not result in closely matching to how LLVM coalesces such comparisons. I'm only lightly familiar with JIT internals or C++ so can't comment further but there are too few diffs and and ARM64 form looked way bigger than I hoped it would be for some reason. Hopefully a better solution is found that brings this area of JIT closer to LLVM/GCC.

Reply:
@pedrobsaila as well..

The JIT already supports creating conditional chains. It was implemented by @a74nh in #79283. It is only enabled on arm64 because only arm64 has conditional compare.

Conditional compare is going to be better than the xor pattern from the example, that's probably why this is a regression on arm64. Also, x64/x86 is getting conditional compare as part of Intel APX, so it is expected that we will enable the same logic for x86/x64 in the future. Given this and the small diffs of this PR I'm not sure the complexity is warranted.

If we want to have this transformation then it should be done by enabling `optOptimizeCompareChainCondBlock` for x86/x64. Then the backend should be taught how to translate this pattern to something more profitable on x86/x64. Currently it only knows how to do that for arm64. The transformation is done by `TryLowerAndOrToCCMP`. x86/x64 should have a similar version that is translated by using `xor` instead. Most likely `optOptimizeCompareChainCondBlock` will have to be restricted on the patterns it allows to combine on x86/x64 since there is no true conditional compare yet.

Reply:
(I removed my comment because I misread this feedback as asking the author rather than a nudge to provide context on disabling on ARM64, though I was hoping to see much larger amount of diffs - this pattern is pretty common after all)

Indeed ccmp is emitted for C: https://godbolt.org/z/srhqPrK4v

Reply:
> Why this check?

yes the ARM support for branchless conditional select is what motivated the if, the xor pattern ends up being more expensive in this case.

> If we want to have this transformation then it should be done by enabling `optOptimizeCompareChainCondBlock` for x86/x64

If I do understand the code well, this would optimize comparison checks for conditional blocks but not for return blocks. Do I need to delete the optimization for return blocks and just do it for conditional ones ?

Reply:
I think it's ok to add some handling for return blocks if the existing handling doesn't cover it. But I think you should transform it into bitwise ops (i.e. `return (x == 0x80) && (y == 0x80)` => `return (x == 0x80) & (y == 0x80)` so that the remaining handling falls out naturally from `TryLowerAndOrToCCMP` (and the x64 specific variant being introduced).